### ISSUE TITLE
PicklingError: Can't pickle <type 'NoneType'>

### ISSUE THREAD

**Issue by basictheprogram** [ansible/ansible#12258]
##### Issue Type:
- Bug Report
##### Ansible Version:

$ ansible --version
ansible 2.0.0 (devel 70677b030b) last updated 2015/09/04 20:34:36 (GMT -500)
  lib/ansible/modules/core: (detached HEAD 2520627fe7) last updated 2015/09/04 20:34:36 (GMT -500)
  lib/ansible/modules/extras: (detached HEAD 5cd0fab9d0) last updated 2015/09/04 20:34:36 (GMT -500)
  config file = /Users/tanner/projects/ansible.git/playbooks.git/ansible.cfg
  configured module search path = None
##### Ansible Configuration:

$ source hacking/env-setup
##### Environment:

OS X Yosemite 10.10.5

$ python -V
Python 2.7.6
##### Summary:

Please summarize your request in this space.  You will earn bonus points for being succinct, but please add enough detail so we can understand the request.  Thanks!
##### Steps To Reproduce:

inventory file vagrant-inventory

```
nagios_server ansible_ssh_host=10.0.1.32
```

nagios-server.yml

```

---
# file: nagios_server.yml
- hosts: nagios_server
  remote_user: vagrant
  vars:
    hostname: nagios3
  roles:
    - { role: bootstrap, sudo: yes }
```

```

---
# tasks file for roles/bootstrap
- name: set hostname
  hostname: name={{ hostname }}
  tags: bootstrap
```
##### Expected Results:

Expect the playbook to complete like it does with 1.9 release.
##### Actual Results:

```
$ ansible-playbook -i vagrant-inventory  nagios-server.yml -vvvv
Using /Users/tanner/projects/ansible.git/playbooks.git/ansible.cfg as config file
1 plays in nagios-server.yml
Loaded callback default of type stdout, v2.0

PLAY ***************************************************************************

TASK [setup] *******************************************************************
<10.0.1.32> ESTABLISH SSH CONNECTION FOR USER: vagrant
<10.0.1.32> EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=600s -o ForwardAgent=yes -o ControlPath="/Users/tanner/.ansible/cp/ansible-ssh-%h-%p-%r" -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=10 10.0.1.32 LANG=C LC_MESSAGES=C LC_CTYPE=C /usr/bin/python
ok: [nagios_server]

TASK [bootstrap : set hostname] ************************************************
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/queues.py", line 268, in _feed
    send(obj)
PicklingError: Can't pickle <type 'NoneType'>: attribute lookup __builtin__.NoneType failed
```

**Comment by basictheprogram** [comment]
The problem isn't in the set host task.

```
---
# tasks file for roles/bootstrap

#- name: set hostname
#  hostname: name={{ hostname }}
#  tags: bootstrap

- name: fix up /etc/hosts
  lineinfile: >
      dest=/etc/hosts 
      regexp='.*{{ hostname }}$'
      line='{{ ansible_default_ipv4.address }}\t\t{{ fqdn }} {{ hostname }}' 
      state=present
  tags: bootstrap
```

```
$ ansible-playbook -i vagrant-inventory  nagios-server.yml

PLAY ***************************************************************************
TASK [setup] *******************************************************************
ok: [nagios_server]
TASK [bootstrap : fix up /etc/hosts] *******************************************
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/queues.py", line 268, in _feed
    send(obj)
PicklingError: Can't pickle <type 'NoneType'>: attribute lookup __builtin__.NoneType failed
```

**Comment by pwnall** [comment]
I'm getting the same stack trace on a completely different task. 2.0 used to work for me, so the change that introduced this must have happened in the last couple of weeks.

```
TASK [os_package_update : update dnf packages] *********************************
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/queues.py", line 268, in _feed
    send(obj)
PicklingError: Can't pickle <type 'NoneType'>: attribute lookup __builtin__.NoneType failed
```

**Comment by basictheprogram** [comment]
In your ansible what's your sftp_batch_mode?

```
# if False, sftp will not use batch mode to transfer files. This may cause some
# types of file transfer failures impossible to catch however, and should
# only be disabled if your sftp version has problems with batch mode
#sftp_batch_mode = False
```

Not an expert, but I'll try some git bisect and see if I can narrow it down.

**Comment by pwnall** [comment]
I did not touch `sftp_batch_mode`. I am just about to start bisecting too.

**Comment by max-b** [comment]
I just ran into this. It looks like v2.0.0-0.1.alpha1 did not cause this error for me, though v2.0.0-0.2.alpha2 did.

**Comment by pwnall** [comment]
If you need a solution right now, https://github.com/ansible/ansible/commit/2d251cba452243888c717b8846ee3ac7387839ab still works for me. I'm down to a few commits in my bisect.

**Comment by pwnall** [comment]
Done bisecting, 87f75a50adc1988cfccd9e1e442b324d8c1fd8c9 broke things. It looks like fancy Python :cry:

**Comment by pwnall** [comment]
The root cause is the `UnsafeProxy` in 87f75a50adc1988cfccd9e1e442b324d8c1fd8c9. If I replace `host_facts[k] = UnsafeProxy(host_facts[k])` with `host_facts[k] = host_facts[k]` everything works.

The UnsafeProxy implementation is really sketchy. It seems like it was copy-pasted from the article that it references, but the author who copy-pasted it ignored the bugs in the comments, including an obvious typo (`__rfloorfiv__` instead of `__rfloordiv__`).

**Comment by jimi-c** [comment]
Ahh interesting, yes I missed comments regarding those typos/bugs. Thanks for catching that! I'll get this merged in, and also have a similar (but slightly different) fix for #12260.

**Comment by jimi-c** [comment]
# Closing This Ticket

Hi!

We believe the above commit should resolve this problem for you. This will also be included in the next major release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
